File: app/src/main/java/com/renatsayf/stockinsider/network/DocumAdapter.java
package com.renatsayf.stockinsider.network;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

public class DocumAdapter implements Converter<ResponseBody, Document>
{
    public static final Converter.Factory FACTORY = new Converter.Factory(){
        @Override
        public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
            if (type == Document.class) return new DocumAdapter();

            return super.responseBodyConverter(type, annotations, retrofit);
        }
    };

    @Override
    public Document convert(ResponseBody value) throws IOException {
        return Jsoup.parse(value.toString());
    }
}

File: app/src/main/java/com/renatsayf/stockinsider/network/DocumAdapter.kt
package com.renatsayf.stockinsider.network

import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import retrofit2.Converter
import retrofit2.Retrofit
import java.io.IOException
import java.lang.reflect.Type

class DocumAdapter : Converter<ResponseBody, Document>
{
    @Throws(IOException::class)
    override fun convert(value : ResponseBody) : Document
    {
        return Jsoup.parse(value.string())
    }

    companion object
    {
        val FACTORY : Converter.Factory = object : Converter.Factory()
        {
            override fun responseBodyConverter(type : Type, annotations : Array<Annotation>, retrofit : Retrofit) : Converter<ResponseBody, *>?
            {
                return if (type === Document::class.java) DocumAdapter()
                else super.responseBodyConverter(type, annotations, retrofit)
            }
        }
    }
}